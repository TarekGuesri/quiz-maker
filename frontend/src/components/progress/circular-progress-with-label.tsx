import React from "react";
import CircularProgress, {
  CircularProgressProps,
} from "@mui/material/CircularProgress";
import Typography from "@mui/material/Typography";
import Box from "@mui/material/Box";

export const CircularProgressWithLabel = (
  props: CircularProgressProps & { value: number },
) => {
  return (
    <Box sx={{ position: "relative", display: "inline-flex" }}>
      <CircularProgress
        variant="determinate"
        size={160}
        sx={{
          color: (theme) =>
            theme.palette.mode === "light"
              ? theme.palette.primary.light
              : theme.palette.primary.dark,
        }}
        {...props}
        color="secondary"
        value={100}
      />
      <CircularProgress
        disableShrink
        sx={{
          color: (theme) =>
            theme.palette.mode === "light"
              ? theme.palette.primary.dark
              : theme.palette.primary.light,
          animationDuration: "550ms",
          position: "absolute",
          left: 0,
        }}
        variant="determinate"
        size={160}
        {...props}
      />
      <Box
        sx={{
          top: 0,
          left: 0,
          bottom: 0,
          right: 0,
          position: "absolute",
          display: "flex",
          alignItems: "center",
          justifyContent: "center",
        }}
      >
        <Typography
          variant="h4"
          component="div"
          color="text.secondary"
        >{`${Math.round(props.value)}%`}</Typography>
      </Box>
    </Box>
  );
};
