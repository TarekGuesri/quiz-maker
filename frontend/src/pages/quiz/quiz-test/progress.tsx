import React from "react";
import LinearProgress from "@mui/material/LinearProgress";
import Typography from "@mui/material/Typography";
import Box from "@mui/material/Box";

interface ProgressProps {
  currentQuestion: number;
  totalQuestions: number;
}

export const Progress: React.FC<ProgressProps> = ({
  currentQuestion,
  totalQuestions,
}) => {
  const progress = (100 * currentQuestion) / totalQuestions;

  return (
    <>
      <Typography mb={2}>
        Question {`${currentQuestion}/${totalQuestions}`}
      </Typography>
      <Box mx="auto" sx={{ width: "100%" }} mb={4}>
        <LinearProgress variant="determinate" value={progress} />
      </Box>
    </>
  );
};
