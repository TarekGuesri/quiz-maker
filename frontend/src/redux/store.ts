import { Action, ThunkAction, configureStore } from "@reduxjs/toolkit";

import ui from "./ui/ui-slice";

export const store = configureStore({
  reducer: {
    ui,
  },
  devTools: true,
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
