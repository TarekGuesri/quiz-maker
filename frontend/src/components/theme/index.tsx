import React, { FC } from 'react';
import CssBaseline from '@mui/material/CssBaseline';
import PropTypes from 'prop-types';
import { createTheme, ThemeProvider } from '@mui/material/styles';
import { useAppSelector } from 'src/redux/hooks';
import varelaRound from 'src/assets/fonts/VarelaRound-Regular.ttf';

export const Theme: FC = (props) => {
  const { darkMode } = useAppSelector((state) => state.ui);

  const theme = createTheme({
    palette: {
      mode: darkMode ? 'dark' : 'light',
      background: {
        default: darkMode ? '#333' : '#f7f7f7',
        paper: darkMode ? '#333' : '#fbfbfb',
      },
      primary: {
        main: '#8e49de',
        // contrastText:
      },
      secondary: {
        main: '#f9de4a',
      },
      contrastThreshold: 3,
      tonalOffset: 0.2,
    },
    typography: {
      fontFamily: 'Varela Round',
    },
    components: {
      MuiCssBaseline: {
        styleOverrides: `
          @font-face {
            font-family: 'Varela Round';
            font-style: normal;
            font-display: swap;
            font-weight: 400;
            src: url(${varelaRound}) format('truetype');
        `,
      },
    },
  });

  return (
    <ThemeProvider theme={theme}>
      {props.children}
      <CssBaseline />
    </ThemeProvider>
  );
};

Theme.propTypes = {
  children: PropTypes.node.isRequired,
};
