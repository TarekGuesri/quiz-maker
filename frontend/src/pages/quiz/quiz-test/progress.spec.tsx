import React from "react";
import { cleanup, render, screen } from "@testing-library/react";
import { store } from "src/redux/store";
import { Provider } from "react-redux";
import { Theme } from "src/components/theme";
import { Progress } from "./progress";

describe("src/pages/quiz/quiz-test/progress.tsx", () => {
  afterEach(() => cleanup());

  test("Render Progress component", () => {
    const renderedComponent = render(
      <Provider store={store}>
        <Theme>
          <Progress currentQuestion={2} totalQuestions={4} />
        </Theme>
      </Provider>,
    );

    expect(renderedComponent.container).toMatchSnapshot();
  });

  test("progress values", () => {
    const test = render(
      <Provider store={store}>
        <Theme>
          <Progress currentQuestion={2} totalQuestions={4} />
        </Theme>
      </Provider>,
    );

    const progressText = screen.getByTestId("progress-text");
    expect(progressText.textContent).toContain("2/4");

    const progressBar = screen.getByTestId("progress-bar");
    expect(progressBar.getAttribute("aria-valuenow")).toBe("50");

    test.rerender(
      <Provider store={store}>
        <Theme>
          <Progress currentQuestion={3} totalQuestions={4} />
        </Theme>
      </Provider>,
    );

    expect(progressText.textContent).toContain("3/4");
    expect(progressBar.getAttribute("aria-valuenow")).toBe("75");
  });
});
