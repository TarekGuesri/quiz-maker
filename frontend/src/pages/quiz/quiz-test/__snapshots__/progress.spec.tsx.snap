// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/pages/quiz/quiz-test/progress.tsx Render Progress component 1`] = `
<div>
  <p
    class="MuiTypography-root MuiTypography-body1 css-j4dx6n-MuiTypography-root"
    data-testid="progress-text"
  >
    Question 
    2/4
  </p>
  <div
    class="MuiBox-root css-1rnc7yp"
  >
    <span
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="50"
      class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate css-1a1si9b-MuiLinearProgress-root"
      data-testid="progress-bar"
      role="progressbar"
    >
      <span
        class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate css-2pg427-MuiLinearProgress-bar1"
        style="transform: translateX(-50%);"
      />
    </span>
  </div>
</div>
`;
